eth_utils/abi.py

```diff
@@ -3,10 +3,37 @@
 from .crypto import keccak
 
 
+def collapse_tuple_type(abi):
+    """Converts a tuple from a dict to a parenthesized list of its types.
+
+    >>> collapse_tuple_type(
+    ...     {
+    ...         'components': [
+    ...             {'name': 'anAddress', 'type': 'address'},
+    ...             {'name': 'anInt', 'type': 'uint256'},
+    ...             {'name': 'someBytes', 'type': 'bytes'},
+    ...         ],
+    ...         'type': 'tuple',
+    ...     }
+    ... )
+    '(address,uint256,bytes)'
+    """
+    if abi["type"] != "tuple":
+        return abi["type"]
+
+    component_types = [
+        collapse_tuple_type(component) for component in abi["components"]
+    ]
+
+    return "(" + ",".join(component_types) + ")"
+
+
 def _abi_to_signature(abi: Dict[str, Any]) -> str:
     function_signature = "{fn_name}({fn_input_types})".format(
         fn_name=abi["name"],
-        fn_input_types=",".join([arg["type"] for arg in abi.get("inputs", [])]),
+        fn_input_types=",".join(
+            [collapse_tuple_type(abi_input) for abi_input in abi["inputs"]]
```

Comment:
This used to handle the case where the `"inputs"` key wasn't present. I'm not sure we can drop that behavior.

Related: it looks like we should add a test where the inputs key is missing.

Reply:
Also, I had to read the code twice because I was surprised that all the abi args were being treated as tuples.

_Just a suggestion:_ Another alternative is to use [formatter applicators](https://eth-utils.readthedocs.io/en/latest/utilities.html#apply-one-of-formatters-condition-formatter-pairs-value-new-value):

```py
def is_tuple(abi):
  return abi["type"] == "tuple"

type_extractor = apply_one_of_formatters((
  (is_tuple, collapse_tuple_type),
  (complement(is_tuple), itemgetter("type")),
))

...

        fn_input_types=",".join(
            [type_extractor(abi_input) for abi_input in abi.get("inputs", [])]
```

This provides a natural upgrade path as other special types are added. 

Also, the `collapse_tuple_type` could validate/require that the type be `'tuple'`.

Reply:
> This used to handle the case where the `"inputs"` key wasn't present. I'm not sure we can drop that behavior.
> 
> Related: it looks like we should add a test where the inputs key is missing.

Great catch, thank you for your scrutiny.